cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go

```diff
@@ -92,45 +96,145 @@ func (c *machineController) findMachineSet(machine *v1alpha1.Machine) (*v1alpha1
+// Run starts shared informers and waits for the informer cache to
+// synchronize.
 func (c *machineController) run(stopCh <-chan struct{}) error {
-	c.informerFactory.Start(stopCh)
+	c.kubeInformerFactory.Start(stopCh)
+	c.clusterInformerFactory.Start(stopCh)
+
+	glog.V(4).Infof("waiting for caches to sync")
+	if !cache.WaitForCacheSync(stopCh,
+		c.nodeInformer.HasSynced,
+		c.machineInformer.Informer().HasSynced,
+		c.machineSetInformer.Informer().HasSynced) {
+		return fmt.Errorf("syncing caches failed")
+	}
+
+	return nil
+}
 
-	glog.V(4).Infof("waiting for machine cache to sync")
-	if !cache.WaitForCacheSync(stopCh, c.machineInformer.Informer().HasSynced) {
-		return fmt.Errorf("cannot sync machine cache")
+// findMachineByNodeProviderID lookups a Node by node.Spec.ProviderID.
+// If the node object is found it looks for a "machine" annotation on
+// the found node to find the underlying machine object. Returns nil
+// if either the Node by node.Spec.ProviderID cannot be found or if
+// the node has no "machine" annotation. A DeepCopy() of the object is
+// returned on success.
+func (c *machineController) findMachineByNodeProviderID(node *apiv1.Node) (*v1alpha1.Machine, error) {
+	objs, err := c.nodeInformer.GetIndexer().ByIndex(nodeProviderIDIndex, node.Spec.ProviderID)
+	if err != nil {
+		return nil, err
+	}
+	if len(objs) != 1 {
+		return nil, nil
 	}
 
-	glog.V(4).Infof("waiting for machineset cache to sync")
-	if !cache.WaitForCacheSync(stopCh, c.machineSetInformer.Informer().HasSynced) {
-		return fmt.Errorf("cannot sync machineset cache")
+	node, ok := objs[0].(*apiv1.Node)
+	if !ok {
+		return nil, fmt.Errorf("internal error; unexpected type %T", node)
 	}
 
-	return nil
+	machineName, found := node.Annotations["machine"]
+	if !found {
+		return nil, nil
+	}
+
+	return c.findMachine(machineName)
 }
 
-func newMachineController(factory informers.SharedInformerFactory) (*machineController, error) {
-	machineInformer := factory.Cluster().V1alpha1().Machines()
-	machineSetInformer := factory.Cluster().V1alpha1().MachineSets()
+// findNodeByNodeName find the Node object keyed by node.Name. Returns
+// nil if it cannot be found. A DeepCopy() of the object is returned
+// on success.
+func (c *machineController) findNodeByNodeName(name string) (*apiv1.Node, error) {
+	item, exists, err := c.nodeInformer.GetIndexer().GetByKey(name)
+	if err != nil {
+		return nil, err
+	}
+
+	if !exists {
+		return nil, nil
+	}
+
+	node, ok := item.(*apiv1.Node)
+	if !ok {
+		return nil, fmt.Errorf("internal error; unexpected type %T", node)
+	}
+
+	return node.DeepCopy(), nil
+}
+
+// MachinesInMachineSet returns all the machines that belong to
+// machineSet. For each machine in the set a DeepCopy() of the object
+// is returned.
+func (c *machineController) MachinesInMachineSet(machineSet *v1alpha1.MachineSet) ([]*v1alpha1.Machine, error) {
+	machines, err := c.machineInformer.Lister().Machines(machineSet.Namespace).List(labels.Everything())
```

Comment:
This is fine with not so many machines. Though, with hundreds of machines in the same namespace (which would be insine but still) this can be slow. What's the benefit of listing all instead of listing all machines that has labels matched by the machineset's label selector?

Reply:
Done.